pkg/icingaredis/heartbeat.go

```diff
@@ -12,79 +13,100 @@ import (
 	"time"
 )
 
+// timeout defines how long a heartbeat may be absent if a heartbeat has already been received.
+// After this time, a heartbeat loss is propagated.
 var timeout = 60 * time.Second
 
+// Heartbeat periodically reads heartbeats from a Redis stream and signals in Beat channels when they are received.
+// Also signals on if the heartbeat is Lost.
 type Heartbeat struct {
-	ctx    context.Context
-	cancel context.CancelFunc
-	client *Client
-	logger *zap.SugaredLogger
-	active bool
-	beat   chan v1.StatsMessage
-	lost   chan struct{}
-	done   chan struct{}
-	mu     *sync.Mutex
-	err    error
+	active    bool
+	beat      *com.Cond
+	cancel    context.CancelFunc
+	client    *Client
+	done      chan struct{}
+	errMu     *sync.Mutex
+	err       error
+	logger    *zap.SugaredLogger
+	lost      *com.Cond
+	message   *v1.StatsMessage
+	messageMu *sync.Mutex
```

Comment:
Given that now everything operates in `Heartbeat` pointers, there is no need for the two mutexes to be pointers. Then their zero value is a fine default and they even don't have to be initialized explicitly.